cmd/plugins/balloons/policy/balloons-policy.go

```diff
@@ -1462,13 +1466,91 @@ func (p *balloons) pinCpuMem(c cache.Container, cpus cpuset.CPUSet, mems idset.I
 	if p.bpoptions.PinMemory == nil || *p.bpoptions.PinMemory {
 		if c.PreserveMemoryResources() {
 			log.Debug("  - preserving %s pinning to memory %q", c.PrettyName, c.GetCpusetMems())
+			preserveMems, err := parseIDSet(c.GetCpusetMems())
+			if err != nil {
+				log.Error("failed to parse CpusetMems: %v", err)
+			} else {
+				zone := p.allocMem(c, preserveMems, true)
+				log.Debug("  - allocated preserved memory %s", c.PrettyName, zone)
+				c.SetCpusetMems(zone.MemsetString())
+			}
 		} else {
 			log.Debug("  - pinning %s to memory %s", c.PrettyName(), mems)
-			c.SetCpusetMems(mems.String())
+			zone := p.allocMem(c, mems, false)
+			log.Debug("  - allocated memory %s", c.PrettyName, zone)
+			c.SetCpusetMems(zone.MemsetString())
 		}
 	}
 }
 
+func (p *balloons) allocMem(c cache.Container, mems idset.IDSet, preserve bool) libmem.NodeMask {
+	var (
+		amount  = getMemoryLimit(c)
+		nodes   = libmem.NewNodeMask(mems.Members()...)
+		req     *libmem.Request
+		zone    libmem.NodeMask
+		updates map[string]libmem.NodeMask
+		err     error
+	)
+
+	if _, ok := p.memAllocator.AssignedZone(c.GetID()); !ok {
+		if c.PreserveMemoryResources() {
+			req = libmem.PreservedContainer(
+				c.GetID(),
+				c.PrettyName(),
+				amount,
+				nodes,
+			)
+		} else {
+			req = libmem.Container(
+				c.GetID(),
+				c.PrettyName(),
+				string(c.GetQOSClass()),
+				amount,
+				nodes,
+			)
+		}
+		zone, updates, err = p.memAllocator.Allocate(req)
+	} else {
+		zone, updates, err = p.memAllocator.Realloc(c.GetID(), nodes, 0)
```

Comment:
A call to Realloc(_, _, **0**) can never change allocation, as memory types mask contains no allowed types. Is the idea to hand control (without types?) to custom expanders? Didn't quite understand this...

Reply:
No... I think I changed the semantics somewhere at some point and the idea was that `libmem.Realloc(id, nodes, 0)` would be equivalent to `libmem(id, nodes, libmem.ZoneType(nodes))`... but either something fell through the cracks somewhere or I changed my mind for some reason. I'll take a look at which one it was...

Reply:
Update with those semantics and added a test case for it. But I kept those as WiP for now. Need to think a bit more about whether these semantics make the most sense, before squashing those commits.

Reply:
The only alternative semantics that I could come up with would be returning an error when trying to reallocate memory with all nodes masked out. In that case `libmem.Realloc(id, nodes, -1)` could be used for reallocating on any possible node.

I could imagine the mask "accidentally" ending up to 0 due to calculating intersections of memory preferences (from annotations and balloontype configurations, for instance), but even in that case the responsibility of finding and handling the issue would be most natural on the caller side of Realloc. And therefore I do not see it necessary to get an error from Realloc, if called with 0.

That is, my gut feeling is almost numb. Without any thinking my personal initial choice might have been -1=any and 0=error, but I like your implementation better, because 0=any scales nicely to more than 64 nodes case. :D